### PR TITLE
[GENX] Initial GENX dialect implementation

### PR DESCRIPTION
**Pull request by etiotto**
Introduce the GENX dialect.

Note: the `libGenISAIntrinsics.a` should be updated in a separate PR. Currently that library has not been built with the same LLVM level we use in this PR.

### CODE REVIEW

File: mlir/include/mlir/Dialect/LLVMIR/GENXDialect.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+#include "mlir/Dialect/LLVMIR/GENXOpsDialect.h.inc"
+#include "mlir/Dialect/LLVMIR/GENXOpsEnums.h.inc"
+#define GET_ATTRDEF_CLASSES
+#include "mlir/Dialect/LLVMIR/GENXOpsAttributes.h.inc"
+#define GET_TYPEDEF_CLASSES
+#include "mlir/Dialect/LLVMIR/GENXOpsTypes.h.inc"
+#define GET_OP_CLASSES
+#include "mlir/Dialect/LLVMIR/GENXOps.h.inc"
+
+#endif // MLIR_DIALECT_LLVMIR_GENXDIALECT_H_
```
Comment (whitneywhtsang) [inline mlir/include/mlir/Dialect/LLVMIR/GENXDialect.h:52]:
```suggestion
#endif // MLIR_DIALECT_LLVMIR_GENXDIALECT_H_

```

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+class GENX_DeviceFunctionOp<string mnemonic, string fnName, 
+                            int arg, list<Trait> traits = []> 
+  : GENX_Op<mnemonic, !listconcat(traits, [Pure])>, 
+    Results<(outs LLVM_Type:$res)>, Arguments<(ins)> {
+  string llvmBuilder = ""
+    # "  llvm::Type *retType = builder.getInt64Ty();\n"
+    # "  llvm::Type *argType = builder.getInt32Ty();\n"
+    # "  llvm::Value *arg = llvm::ConstantInt::get(argType, " # arg # ");\n"
+    # "  $res = createDeviceFunctionCall(builder, \"" # fnName # "\",\n"
+    # "                             retType, { argType }, { arg });";
```
Comment (whitneywhtsang) [inline mlir/include/mlir/Dialect/LLVMIR/GENXOps.td:77]: would something like this work?
```suggestion
  string llvmBuilder = [{
     llvm::Type *retType = builder.getInt64Ty();
     llvm::Type *argType = builder.getInt32Ty();
     llvm::Value *arg = llvm::ConstantInt::get({argType},  $arg);
     $res = createDeviceFunctionCall(builder, $fnName, retType, { argType }, { arg });";
  }];
```

File: mlir/include/mlir/InitAllDialects.h [hunk 3]
```diff
@@ -117,6 +118,7 @@ inline void registerAllDialects(DialectRegistry &registry) {
                   arm_sve::ArmSVEDialect,
                   vector::VectorDialect,
                   NVVM::NVVMDialect,
+                  GENX::GENXDialect,
```
Comment (whitneywhtsang) [inline mlir/include/mlir/InitAllDialects.h:121]:
```suggestion
                  GENX::GENXDialect,
                  NVVM::NVVMDialect,
```

File: mlir/include/mlir/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+/// given registry;
+void registerGENXDialectTranslation(DialectRegistry &registry);
+
+/// Register the GENX dialect and the translation from it in the registry
+/// associated with the given context.
+void registerGENXDialectTranslation(MLIRContext &context);
+
+} // namespace mlir
+
+#endif // MLIR_TARGET_LLVMIR_DIALECT_GENX_GENXTOLLVMIRTRANSLATION_H
```
Comment (whitneywhtsang) [inline mlir/include/mlir/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.h:31]:
```suggestion
#endif // MLIR_TARGET_LLVMIR_DIALECT_GENX_GENXTOLLVMIRTRANSLATION_H

```

File: mlir/lib/Dialect/LLVMIR/CMakeLists.txt [hunk 5]
```diff
@@ -80,3 +80,26 @@ add_mlir_dialect_library(MLIRROCDLDialect
...
+  LINK_COMPONENTS
+  AsmParser
+  Core
+
+  LINK_LIBS PUBLIC
+  MLIRIR
+  MLIRSideEffectInterfaces
+  MLIRVectorToLLVM
+  MLIRLLVMDialect
+  )
```
Comment (whitneywhtsang) [inline mlir/lib/Dialect/LLVMIR/CMakeLists.txt:105]:
```suggestion
  )
  
```

File: mlir/include/mlir/Dialect/LLVMIR/GENXDialect.h (hunk 1 above)
Comment (etiotto) [reply mlir/include/mlir/Dialect/LLVMIR/GENXDialect.h]: done

File: mlir/include/mlir/Dialect/LLVMIR/GENXOps.td (hunk 2 above)
Comment (etiotto) [reply mlir/include/mlir/Dialect/LLVMIR/GENXOps.td:77]: No has to be a string

File: mlir/include/mlir/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.h (hunk 4 above)
Comment (etiotto) [reply mlir/include/mlir/Target/LLVMIR/Dialect/GENX/GENXToLLVMIRTranslation.h]: done

File: mlir/lib/Dialect/LLVMIR/CMakeLists.txt (hunk 5 above)
Comment (etiotto) [reply mlir/lib/Dialect/LLVMIR/CMakeLists.txt]: done

File: mlir/include/mlir/InitAllDialects.h (hunk 3 above)
Comment (etiotto) [reply mlir/include/mlir/InitAllDialects.h]: done

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**